pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/SegmentPreprocessThrottler.java

```diff
@@ -0,0 +1,212 @@
+    int maxPreprocessConcurrency = Integer.parseInt(maxParallelSegmentPreprocessesStr);
+
+    if (maxPreprocessConcurrency == _maxPreprocessConcurrency) {
+      LOGGER.info("No ZK update for maxPreprocessConcurrency {}, total permits: {}", _maxPreprocessConcurrency,
+          totalPermits());
+      return;
+    }
+
+    if (maxPreprocessConcurrency <= 0) {
+      LOGGER.warn("Invalid maxPreprocessConcurrency set: {}, not making change, fix config and try again",
+          maxPreprocessConcurrency);
+      return;
+    }
+
+    LOGGER.info("Updated maxPreprocessConcurrency from: {} to: {}", _maxPreprocessConcurrency,
+        maxPreprocessConcurrency);
+    _maxPreprocessConcurrency = maxPreprocessConcurrency;
+
+    if (_relaxThrottling) {
+      LOGGER.warn("Reset throttling hasn't been called yet, not updating the permits with maxPreprocessConcurrency");
+      return;
+    }
+    _semaphore.setPermits(_maxPreprocessConcurrency);
+  }
+
+  private void handleMaxPreprocessConcurrencyBeforeServingQueriesChange(Set<String> changedConfigs,
+      Map<String, String> clusterConfigs) {
+    if (!changedConfigs.contains(
+        CommonConstants.Helix.CONFIG_OF_MAX_SEGMENT_PREPROCESS_PARALLELISM_BEFORE_SERVING_QUERIES)) {
+      LOGGER.info("changedConfigs list indicates maxPreprocessConcurrencyBeforeServingQueries was not updated, "
+          + "skipping updates");
+      return;
+    }
+
+    String configName = CommonConstants.Helix.CONFIG_OF_MAX_SEGMENT_PREPROCESS_PARALLELISM_BEFORE_SERVING_QUERIES;
+    String defaultConfigValue = CommonConstants.Helix.DEFAULT_MAX_SEGMENT_PREPROCESS_PARALLELISM_BEFORE_SERVING_QUERIES;
+    String maxParallelSegmentPreprocessesBeforeServingQueriesStr =
+        clusterConfigs.getOrDefault(configName, defaultConfigValue);
+    int maxPreprocessConcurrencyBeforeServingQueries =
+        Integer.parseInt(maxParallelSegmentPreprocessesBeforeServingQueriesStr);
+
+    if (maxPreprocessConcurrencyBeforeServingQueries == _maxPreprocessConcurrencyBeforeServingQueries) {
+      LOGGER.info("No ZK update for maxPreprocessConcurrencyBeforeServingQueries {}, total permits: {}",
+          _maxPreprocessConcurrencyBeforeServingQueries, totalPermits());
+      return;
+    }
+
+    if (maxPreprocessConcurrencyBeforeServingQueries <= 0) {
+      LOGGER.warn("Invalid maxPreprocessConcurrencyBeforeServingQueries set: {}, not making change, fix config "
+              + "and try again", maxPreprocessConcurrencyBeforeServingQueries);
+      return;
+    }
+
+    LOGGER.info("Updated maxPreprocessConcurrencyBeforeServingQueries from: {} to: {}",
+        _maxPreprocessConcurrencyBeforeServingQueries, maxPreprocessConcurrencyBeforeServingQueries);
+    _maxPreprocessConcurrencyBeforeServingQueries = maxPreprocessConcurrencyBeforeServingQueries;
+    if (_relaxThrottling) {
+      LOGGER.warn("maxPreprocessConcurrencyBeforeServingQueries was updated before reset throttling was called, "
```

Comment:
This shouldn't be a warning. We can log an info if this is changed after serving queries

Reply:
done